jetbrains-core/src/software/aws/toolkits/jetbrains/services/cloudformation/Cloudformation.kt

```diff
@@ -0,0 +1,43 @@
+// Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package software.aws.toolkits.jetbrains.services.cloudformation
+
+import software.amazon.awssdk.services.cloudformation.CloudFormationClient
+import software.amazon.awssdk.services.cloudformation.model.CloudFormationException
+import software.amazon.awssdk.services.cloudformation.model.DescribeStacksRequest
+import software.amazon.awssdk.services.cloudformation.model.StackStatus
+import software.aws.toolkits.resources.message
+import java.util.concurrent.TimeoutException
+
+fun CloudFormationClient.waitForStackDeletionComplete(stackName: String, maxAttempts: Int, delay: Int) {
+    val deletionSuccessStatuses = setOf(StackStatus.DELETE_COMPLETE)
+    val deletionFailedStatuses = setOf(StackStatus.DELETE_FAILED, StackStatus.ROLLBACK_FAILED, StackStatus.CREATE_FAILED, StackStatus.UPDATE_ROLLBACK_IN_PROGRESS, StackStatus.UPDATE_ROLLBACK_FAILED)
+
+    repeat(maxAttempts) {
+        val stacks = try {
+            this.describeStacks(DescribeStacksRequest.builder().stackName(stackName).build()).stacks()
+        } catch (e: CloudFormationException) {
+            if (e.awsErrorDetails().errorCode() == "ValidationError") {
+                return
+            }
+            throw e
+        }
+        if (stacks.size != 1) {
+            throw RuntimeException(message("cloudformation.delete_stack.multiple_stacks_with_same_name"))
+        }
+
+        val targetStack = stacks[0]
+        when (targetStack.stackStatus()) {
```

Comment:
We probably want this whole thing to be a more generic "waiter" function - there's going to be other places where we need to repeat an SDK call until a terminal state is reached. Something like:

```
fun <T> waiter(call: () -> T, successPredicate: (T) -> Boolean, failurePredicate: (T) -> Boolean, attempts: Int, delay: Duration)
```

This would handle all of your `repeat` logic and would match on the state using the predicates.

Take a look at the waiter in the SDK's test utils - `https://github.com/aws/aws-sdk-java-v2/blob/master/test/test-utils/src/main/java/software/amazon/awssdk/testutils/Waiter.java`

Reply:
Good point. That's exactly what I'm thinking now as I'm working on the deploy SAM. It has a lot status change handling. Will rework this for generic waiters.